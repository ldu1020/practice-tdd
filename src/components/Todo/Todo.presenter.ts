class TodoPresenter {
  list: Item[] = [];
  constructor(initialList: Item[]) {
    this.list = [...initialList];
  }

  addItem(item: Item, updateFn?: UpdateFn) {
    this.list = [...this.list, item];
    if (updateFn) updateFn(this.list);
  }

  deleteItem(id: number, updateFn?: UpdateFn) {
    const updated = this.list.filter((li) => li.id !== id);
    this.list = updated;
    if (updateFn) updateFn(this.list);
  }

  checkItem(id: number, updateFn?: UpdateFn) {
    const updated = this.list.map((li) => {
      if (li.id === id) li.checked = !li.checked;
      return li;
    });
    this.list = updated;
    if (updateFn) updateFn(this.list);
  }

  getList() {
    return this.list;
  }

  get isEmpty() {
    return this.list.length === 0;
  }
}

export default TodoPresenter;

export type Item = {
  id: number;
  title: string;
  checked: boolean;
};

export type UpdateFn = (items: Item[]) => void;
