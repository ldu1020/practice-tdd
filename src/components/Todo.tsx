import React from "react";
import { Box } from "@chakra-ui/layout";
import TodoPresenter, { Item } from "./Todo/Todo.presenter";
import TodoAddForm from "./Todo/TodoAddForm/TodoAddForm";
import TodoList from "./Todo/TodoList/TodoList";

interface TodoProps {
  todoPresenter: TodoPresenter;
}

const Todo = ({ todoPresenter }: TodoProps) => {
  const [list, setList] = React.useState(todoPresenter.getList());
  const handleAddTodo = (item: Item) => {
    todoPresenter.addItem(item, setList);
  };

  const handleDeleteTodo = (id: number) => {
    todoPresenter.deleteItem(id, setList);
  };

  const handleCheckTodo = (id: number) => {
    todoPresenter.checkItem(id, setList);
  };

  return (
    <Box w="300px" border="1px">
      <TodoAddForm onAdd={handleAddTodo} />
      <TodoList
        list={list}
        onCheck={handleCheckTodo}
        onDelete={handleDeleteTodo}
      />
    </Box>
  );
};

export default Todo;
