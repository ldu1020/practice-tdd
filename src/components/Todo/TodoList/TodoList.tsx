import React from "react";
import { Flex, List, ListIcon, ListItem } from "@chakra-ui/layout";
import { Item } from "../Todo.presenter";
import { CheckIcon } from "@chakra-ui/icons";
import { Button } from "@chakra-ui/react";

interface TodoListProps {
  list: Item[];
  onCheck: (id: number) => void;
  onDelete: (id: number) => void;
}

const TodoList = ({ list, onCheck, onDelete }: TodoListProps) => {
  return (
    <List>
      {list.map((item) => {
        return (
          <ListItem
            key={item.id}
            display="flex"
            justifyContent="space-between"
            alignItems="center"
            borderBottom="1px"
            p="10px"
            cursor="pointer"
            onClick={() => {
              onCheck(item.id);
            }}
          >
            {item.title}
            <Flex>
              <ListIcon as={CheckIcon} color={item.checked ? "red" : "gray"} />
              <Button
                onClick={() => {
                  onDelete(item.id);
                }}
              >
                Delete
              </Button>
            </Flex>
          </ListItem>
        );
      })}
    </List>
  );
};

export default TodoList;
