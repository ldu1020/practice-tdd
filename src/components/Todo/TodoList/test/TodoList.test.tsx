import { render } from "@testing-library/react";
import { screen } from "@testing-library/dom";
import renderer from "react-test-renderer";
import { Item } from "../../Todo.presenter";
import TodoList from "../TodoList";
import userEvent from "@testing-library/user-event";

describe("TodoList", () => {
  let list: Item[];
  let item1: Item;
  let item2: Item;
  let item3: Item;

  let onCheck: (id: number) => void;
  let onDelete: (id: number) => void;
  let setUp = () =>
    render(<TodoList list={list} onCheck={onCheck} onDelete={onDelete} />);

  beforeEach(() => {
    item1 = { id: 1, title: "first", checked: false };
    item2 = { id: 2, title: "two", checked: false };
    item3 = { id: 3, title: "three", checked: false };
    list = [item1, item2, item3];

    onCheck = jest.fn();
    onDelete = jest.fn();
  });

  it("renders", () => {
    const component = renderer.create(
      <TodoList list={list} onCheck={onCheck} onDelete={onDelete} />
    );
    expect(component.toJSON()).toMatchSnapshot();
  });

  it("calls 'onDelete' when click 'Delete' button with id of Item", () => {
    setUp();
    const btns = screen.getAllByText("Delete");
    btns.forEach((btn, idx) => {
      userEvent.click(btn);
      expect(onDelete).toHaveBeenCalledWith(list[idx].id);
    });
  });

  it("calls 'onCheck' when click 'Item' button with id of Item", () => {
    setUp();
    list.forEach((item) => {
      const itemDom = screen.getByText(item.title);
      userEvent.click(itemDom);
      expect(onCheck).toHaveBeenCalledWith(item.id);
    });
  });
});
