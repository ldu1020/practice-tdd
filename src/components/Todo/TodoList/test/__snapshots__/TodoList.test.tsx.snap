// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoList renders 1`] = `
<ul
  className="css-155za0w"
  role="list"
>
  <li
    className="css-1xkergn"
    onClick={[Function]}
  >
    first
    <div
      className="css-k008qs"
    >
      <svg
        className="chakra-icon chakra-icon css-1kgncjz"
        focusable={false}
        role="presentation"
        viewBox="0 0 14 14"
      >
        <g
          fill="currentColor"
        >
          <polygon
            points="5.5 11.9993304 14 3.49933039 12.5 2 5.5 8.99933039 1.5 4.9968652 0 6.49933039"
          />
        </g>
      </svg>
      <button
        className="chakra-button css-5c2176"
        onClick={[Function]}
        type="button"
      >
        Delete
      </button>
    </div>
  </li>
  <li
    className="css-1xkergn"
    onClick={[Function]}
  >
    two
    <div
      className="css-k008qs"
    >
      <svg
        className="chakra-icon chakra-icon css-1kgncjz"
        focusable={false}
        role="presentation"
        viewBox="0 0 14 14"
      >
        <g
          fill="currentColor"
        >
          <polygon
            points="5.5 11.9993304 14 3.49933039 12.5 2 5.5 8.99933039 1.5 4.9968652 0 6.49933039"
          />
        </g>
      </svg>
      <button
        className="chakra-button css-5c2176"
        onClick={[Function]}
        type="button"
      >
        Delete
      </button>
    </div>
  </li>
  <li
    className="css-1xkergn"
    onClick={[Function]}
  >
    three
    <div
      className="css-k008qs"
    >
      <svg
        className="chakra-icon chakra-icon css-1kgncjz"
        focusable={false}
        role="presentation"
        viewBox="0 0 14 14"
      >
        <g
          fill="currentColor"
        >
          <polygon
            points="5.5 11.9993304 14 3.49933039 12.5 2 5.5 8.99933039 1.5 4.9968652 0 6.49933039"
          />
        </g>
      </svg>
      <button
        className="chakra-button css-5c2176"
        onClick={[Function]}
        type="button"
      >
        Delete
      </button>
    </div>
  </li>
</ul>
`;
